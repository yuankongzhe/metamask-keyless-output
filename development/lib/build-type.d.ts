import { Infer, Struct } from 'superstruct';

export type Unique<Element extends Struct<any>> = (
  struct: Struct<Infer<Element>[], Infer<Element>>,
  eq?: (a: Infer<Element>, b: Infer<Element>) => boolean,
) => Struct<Infer<Element>[], Infer<Element>>;
