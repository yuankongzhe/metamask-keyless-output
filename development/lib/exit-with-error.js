/**
 * Exit the process with an error message.
 *
 * Note that this should be called before the process ends, but it will not
 * itself end the process. This is because the Node.js documentation strongly
 * advises against calling `process.exit` directly.
 *
 * @param {string} errorMessage - The error message that is causing the non-
 *   zero exit code.
 */
function exitWithError(errorMessage) {
  console.error(errorMessage);
  process.exitCode = 1;
  if (process.env.CI) {
    // For some reason, the CI environment does not exit when process.exitCode
    // is set, or at least doesn't terminate immediately. Something may be
    // keeping the event loop from ending. The result is that sometimes jobs
    // fail because of timeout even though none of the individual tests fail.
    // This is a workaround.
    process.exit();
  }
}

module.exports = { exitWithError };
