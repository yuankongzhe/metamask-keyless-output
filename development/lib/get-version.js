const { version: manifestVersion } = require('../../package.json');
const { loadBuildTypesConfig } = require('./build-type');

/**
 * Get the current version of the MetaMask extension. The base manifest version
 * is modified according to the build type and version.
 *
 * The build version is needed because certain build types (such as beta) may
 * be released multiple times during the release process.
 *
 * @param {string} buildType - The build type.
 * @param {number} buildVersion - The build version.
 * @returns {string} The MetaMask extension version.
 */
function getVersion(buildType, buildVersion) {
  return loadBuildTypesConfig().buildTypes[buildType].isPrerelease === true
    ? `${manifestVersion}-${buildType}.${buildVersion}`
    : manifestVersion;
}

module.exports = { getVersion };
