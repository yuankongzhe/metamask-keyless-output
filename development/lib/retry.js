/**
 * Re-runs the given function until it returns a resolved promise or the number
 * of retries is exceeded, whichever comes first (with an optional delay in
 * between retries).
 *
 * @param {object} args - A set of arguments and options.
 * @param {number} args.retries - The maximum number of times to re-run the
 * function on failure.
 * @param {number} [args.delay] - The amount of time (in milliseconds) to wait in
 * between retries. (Default: 0)
 * @param {string} [args.rejectionMessage] - The message for the rejected promise
 * this function will return in the event of failure. (Default: "Retry limit
 * reached")
 * @param {boolean} [args.retryUntilFailure] - Retries until the function fails.
 * @param {Function} functionToRetry - The function that is run and tested for
 * failure.
 * @returns {Promise<* | null | Error>} a promise that either resolves with one of the following:
 * - If successful, resolves with the return value of functionToRetry.
 * - If functionToRetry fails while retryUntilFailure is true, resolves with null.
 * - Otherwise it is rejected with rejectionMessage.
 */
async function retry(
  {
    retries,
    delay = 0,
    rejectionMessage = 'Retry limit reached',
    retryUntilFailure = false,
  },
  functionToRetry,
) {
  let attempts = 0;
  while (attempts <= retries) {
    if (attempts > 0 && delay > 0) {
      await new Promise((resolve) => setTimeout(resolve, delay));
    }

    try {
      const result = await functionToRetry();
      if (!retryUntilFailure) {
        return result;
      }
    } catch (error) {
      if (error.message === "Exited with code '1'") {
        console.log("retry() received: Exited with code '1'");
      } else {
        console.error('error caught in retry():', error);
      }

      if (attempts < retries) {
        console.log('Ready to retry() again');
      }

      if (retryUntilFailure) {
        return null;
      }
    } finally {
      attempts += 1;
    }
  }

  throw new Error(rejectionMessage);
}

module.exports = { retry };
