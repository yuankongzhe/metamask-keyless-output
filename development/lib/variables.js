/* eslint-disable jsdoc/check-tag-names */
const assert = require('assert');

const DeclaredOnly = Symbol(
  'This variable was declared only without being defined',
);

class Variables {
  /**
   * @type {Map<string, unknown | typeof DeclaredOnly>}
   */
  #definitions = new Map();

  /**
   * @param {Iterable<string>} declarations
   */
  constructor(declarations) {
    for (const declaration of declarations) {
      this.#definitions.set(declaration, DeclaredOnly);
    }
  }

  /**
   * @param {string} key - The name of the variable
   * @throws {TypeError} If there is no definition of a variable.
   */
  get(key) {
    const value = this.getMaybe(key);
    assert(
      value !== DeclaredOnly,
      new TypeError(
        `Tried to access a declared, but not defined environmental variable "${key}"
\tWhy am I seeing this: "${key}" is declared in builds.yml, but had no actual value when we tried loading it.
\tHow do I fix this: You could provide a default value for the variable in builds.yml under "env" property and commit to git. For example:
\t\tenv:
\t\t - ${key}: ''`,
      ),
    );
    return value;
  }

  /**
   * Returns a declared, but maybe not defined variable.
   *
   * @param {string} key - The name of the variable
   * @throws {TypeError} If there was no declaration of the variable.
   * @returns The value, or undefined if the variables wasn't defined.
   */
  getMaybe(key) {
    assert(
      this.isDeclared(key),
      new TypeError(
        `Tried to access an environmental variable "${key}" that wasn't declared in builds.yml
\tWhy am I seeing this: We've made use of new variables be explicit to keep track of all of them in one place
\tHow do I fix this: Adding your variable in builds.yml under "env" property and committing to git will fix this`,
      ),
    );
    return this.#definitions.get(key);
  }

  /**
   * Sets one key
   *
   * @overload
   * @param {string} key
   * @param {unknown} value
   * @returns {void}
   */
  /**
   * @overload
   * @param {Record<string, unknown>} records - Key-Value object
   * @returns {void}
   */
  /**
   * @param {string | Record<string, unknown>} keyOrRecord
   * @param {unknown} value
   * @returns {void}
   */
  set(keyOrRecord, value) {
    if (typeof keyOrRecord === 'object') {
      for (const [key, recordValue] of Object.entries(keyOrRecord)) {
        this.set(key, recordValue);
      }
      return;
    }
    const key = keyOrRecord;
    assert(
      this.isDeclared(key),
      `Tried to modify a variable "${key}" that wasn't declared in builds.yml`,
    );
    assert(value !== DeclaredOnly, `Tried to un-define "${key}" variable`);
    this.#definitions.set(key, value);
  }

  isDeclared(key) {
    return this.#definitions.has(key);
  }

  isDefined(key) {
    return (
      this.#definitions.has(key) && this.#definitions.get(key) !== DeclaredOnly
    );
  }

  [Symbol.iterator] = this.declarations;

  *declarations() {
    yield* this.#definitions.keys();
  }

  *definitions() {
    for (const [key, value] of this.#definitions.entries()) {
      if (value !== DeclaredOnly) {
        yield [key, value];
      }
    }
  }
}

module.exports = { Variables };
