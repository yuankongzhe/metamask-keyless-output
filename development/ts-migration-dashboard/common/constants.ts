import path from 'path';

export const ROOT_DIRECTORY_PATH = path.resolve(__dirname, '../../..');
export const PROJECT_DIRECTORY_PATH = path.join(
  ROOT_DIRECTORY_PATH,
  'development',
  'ts-migration-dashboard',
);
export const COMMON_DIRECTORY_PATH = path.join(
  PROJECT_DIRECTORY_PATH,
  'common',
);
export const APP_DIRECTORY_PATH = path.join(PROJECT_DIRECTORY_PATH, 'app');
export const INTERMEDIATE_BUILD_DIRECTORY_PATH = path.join(
  PROJECT_DIRECTORY_PATH,
  'build',
  'intermediate',
);
export const FINAL_BUILD_DIRECTORY_PATH = path.join(
  PROJECT_DIRECTORY_PATH,
  'build',
  'final',
);
export const ENTRYPOINT_PATTERNS = [
  'app/scripts/background',
  'app/scripts/contentscript',
  'app/scripts/disable-console',
  'app/scripts/inpage',
  'app/scripts/phishing-detect',
  'app/scripts/sentry-install',
  'app/scripts/ui',
  'development/build/index',
  '**/*.stories',
  '**/*.test',
];
export const FILES_TO_CONVERT_PATH = path.join(
  PROJECT_DIRECTORY_PATH,
  'files-to-convert.json',
);
