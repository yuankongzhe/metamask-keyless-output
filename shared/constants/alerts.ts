export enum AlertTypes {
  unconnectedAccount = 'unconnectedAccount',
  web3ShimUsage = 'web3ShimUsage',
  invalidCustomNetwork = 'invalidCustomNetwork',
}

/**
 * Alerts that can be enabled or disabled by the user.
 */
export const TOGGLEABLE_ALERT_TYPES = [
  AlertTypes.unconnectedAccount,
  AlertTypes.web3ShimUsage,
];

export enum Web3ShimUsageAlertStates {
  recorded = 1,
  dismissed = 2,
}
