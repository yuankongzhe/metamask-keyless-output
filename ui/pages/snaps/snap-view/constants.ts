export const KeyringSnapRemovalResultStatus = {
  Success: 'success',
  Failed: 'failed',
  None: 'none',
};
