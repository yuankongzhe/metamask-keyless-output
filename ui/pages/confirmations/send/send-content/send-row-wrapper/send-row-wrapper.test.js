import React from 'react';
import configureMockStore from 'redux-mock-store';
import { renderWithProvider } from '../../../../../../test/lib/render-helpers';
import { INSUFFICIENT_FUNDS_ERROR_KEY } from '../../../../../helpers/constants/error-keys';
import SendRowWrapper from '.';

describe('SendContent Component', () => {
  describe('render', () => {
    it('should render with children', () => {
      const props = {
        errorType: 'mockErrorType',
        label: 'mockLabel',
        showError: false,
      };
      const { container } = renderWithProvider(
        <SendRowWrapper {...props}>
          <span>Mock Custom Label Content</span>
          <span>Mock Form Field</span>
        </SendRowWrapper>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render a SendRowErrorMessage with and errorType props if showError is true', () => {
      const mockState = {
        send: {
          currentTransactionUUID: '1-tx',
          draftTransactions: {
            '1-tx': {
              gas: {
                error: INSUFFICIENT_FUNDS_ERROR_KEY,
              },
              amount: {
                error: INSUFFICIENT_FUNDS_ERROR_KEY,
              },
            },
          },
        },
      };

      const mockStore = configureMockStore()(mockState);

      const props = {
        errorType: 'amount',
        label: 'mockLabel',
        showError: true,
      };
      const { container } = renderWithProvider(
        <SendRowWrapper {...props}>
          <span>Mock Form Field</span>
        </SendRowWrapper>,
        mockStore,
      );

      expect(container).toMatchSnapshot();
    });
  });
});
