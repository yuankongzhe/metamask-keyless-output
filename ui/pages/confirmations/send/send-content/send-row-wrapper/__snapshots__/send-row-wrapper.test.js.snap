// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendContent Component render should render a SendRowErrorMessage with and errorType props if showError is true 1`] = `
<div>
  <div
    class="send-v2__form-row"
  >
    <div
      class="send-v2__form-label"
    >
      mockLabel
    </div>
    <div
      class="send-v2__form-field-container"
    >
      <div
        class="send-v2__form-field"
      >
        <span>
          Mock Form Field
        </span>
      </div>
      <div>
        <div
          class="send-v2__error send-v2__error-amount"
        >
          Insufficient funds.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SendContent Component render should render with children 1`] = `
<div>
  <div
    class="send-v2__form-row"
  >
    <div
      class="send-v2__form-label"
    >
      mockLabel
      <span>
        Mock Custom Label Content
      </span>
    </div>
    <div
      class="send-v2__form-field"
    >
      <span>
        Mock Form Field
      </span>
    </div>
  </div>
</div>
`;
